### DISCUSSION TITLE
Does NVDA depend on IAccessible API

### DISCUSSION THREAD

**Discussion by isidorn** [nvaccess/nvda#13784] [Q&A] [answered]
We are having a discussion on the VS Code since we want to improve the screen reader detection.
Currently accessibility clients of Chrome use legacy IAccessible MSAA or the new IAccessible2 UIAutomation.
We would like to drop support for auto detecting screen readers that support the legacy IAccessible MSAA. 

Does NVDA depend on the legacy `IAccessible`API?

Thanks a lot
Isidor

**Comment by Brian1Gaff** [comment]
I think that most screenreaders keep it as a fall back for some purposes, as 
I see it mentioned in log files here and there.
 Brian

-- 
***@***.***
Sent via blueyonder.(Virgin media)
Please address personal E-mail to:-
***@***.***, putting 'Brian Gaff'
in the display name field.
----- Original Message ----- 
From: "Isidor Nikolic" ***@***.***>
To: "nvaccess/nvda" ***@***.***>
Cc: "Subscribed" ***@***.***>
Sent: Thursday, June 09, 2022 3:07 PM
Subject: [nvaccess/nvda] Does NVDA depend on IAccessible API (Issue #13782)

**Comment by isidorn** [comment]
Thanks. But that means that NVDA does use `IAccessible2`?
@feerrenrut can you maybe confirm when you find time. Thanks you

**Comment by lukaszgo1** [comment]
@isidorn  I'm not sure your question can be answered as it stands. You wrote:

> Currently accessibility clients of Chrome use legacy IAccessible MSAA or the new IAccessible2 UIAutomation. 

That is not really true. IAccessible2 and UIAutomation are two separate API's. For more explanation see [this SO question](https://stackoverflow.com/questions/55129774/what-is-the-difference-between-iaccessible-iaccessible2-uiautomation-and-msaa)

Perhaps if you would tell us
1. How exactly you're detecting screen readers now?
2. What you're trying to change and
3. What problem you're trying to solve

we would be able to provide a more complete answer.
If you're asking about IAccessible2, then yes, NVDA, and for that matter all Windows screen readers except Narrator, rely on it to access Electron based applications, as it offers access far superior to what UIA allows.

**Comment by feerrenrut** [comment, marked as answer]
Yes, I can confirm the answer by @lukaszgo1. NVDA uses a mix of (depending on scenario):
- MSAA/IAccessible
- IAccessble2
- UIA (Both batched via [remote operations ](https://github.com/microsoft/Microsoft-UI-UIAutomation/blob/main/README.md) and individual cross process calls)
- Java Access Bridge

Historically for Chromium IAccessible2 has been used (supplementing MSAA/IAccessible), however support for UIA is also available and can be enabled / forced via the advanced settings menu of NVDA. You may have to ensure that the browser is exposing it with command line flags. That said, IAccessible2 based implementation for browsers is more battle tested, essentially the same implementation is used for Firefox, Chrome and Edge. Although there are regular changes, the core implementation has existed for many years.

**Comment by isidorn** [comment]
Thank you very much, this confirms that NVDA uses IAccessible2 for Chromium, which is what we were interested in.
I appreciate your help.
